File: 139.word-break.cpp
/*
 * @lc app=leetcode id=139 lang=cpp
 *
 * [139] Word Break
 */

#include <algorithm>
#include <iterator>
#include <random>
#include <set>
#include <string>
#include <string_view>
#include <vector>
using std::string;
using std::string_view;
using std::vector;

// @lc code=start
class Solution {
  public:
    bool solve(const string_view s, vector<string> &dict) {
        for (const string &word : dict) {
            if (s == word) return true;
        }

        for (const string &word : dict) {
            if (s.starts_with(word)) {
                if (solve(s.substr(word.length()), dict)) { return true; }
            }
        }

        return false;
    }

    bool wordBreak(string s, vector<string> &wordDict) {
        std::set<char> ss(s.begin(), s.end());
        std::set<char> sw;
        for (const string& w : wordDict) {
            sw.insert(w.begin(), w.end());
        }
        if (!std::includes(sw.begin(), sw.end(), ss.begin(), ss.end())) return false;

        std::sort(wordDict.begin(), wordDict.end(), [](const string &str1, const string &str2) { return str1.length() < str2.length(); });
        string_view sv(s);
        return solve(sv, wordDict);
    }
};
// @lc code=end

File: 139.word-break.go
package leetcode

/*
 * @lc app=leetcode id=139 lang=golang
 *
 * [139] Word Break
 */

// @lc code=start

func wordBreak_I(s string, wordDict []string) bool {
	dp := make([]bool, len(s)+1)
	words := make(map[string]bool, len(wordDict))
	for i := range wordDict {
		words[wordDict[i]] = true
	}
	dp[0] = true
	max_word_len := 0
	for i := range wordDict {
		max_word_len = max(max_word_len, len(wordDict[i]))
	}

	// i should start from 1
	// so use i+1
	for i := range s {
		for j := i; j >= max(i+1-max_word_len, 0); j-- {
			if dp[j] && words[s[j:i+1]] {
				dp[i+1] = true
			}
		}
	}

	return dp[len(s)]
}

// @lc code=end
